mm2src/mm2_p2p/src/behaviours/peers_exchange.rs

```diff
@@ -0,0 +1,394 @@
+use futures::StreamExt;
+use futures_ticker::Ticker;
+use libp2p::{multiaddr::Protocol,
+             request_response::{Behaviour as RequestResponse, Config as RequestResponseConfig, Event, HandlerEvent,
+                                InboundFailure, OutboundFailure, ProtocolSupport, ResponseChannel},
+             swarm::{NetworkBehaviour, ToSwarm},
+             Multiaddr, PeerId};
+use log::{info, warn};
+use rand::seq::SliceRandom;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::{collections::{HashMap, HashSet, VecDeque},
+          iter,
+          task::Poll,
+          time::Duration};
+
+use super::request_response::Codec;
+use crate::NetworkInfo;
+// use crate::peer_store::Behaviour as PeerStoreBehaviour;
+
+pub type PeerAddresses = HashSet<Multiaddr>;
+type PeersExchangeCodec = Codec<PeersExchangeProtocol, PeersExchangeRequest, PeersExchangeResponse>;
+
+const DEFAULT_PEERS_NUM: usize = 20;
+const REQUEST_PEERS_INITIAL_DELAY: u64 = 20;
+const REQUEST_PEERS_INTERVAL: u64 = 300;
+const MAX_PEERS: usize = 100;
+
+#[derive(Debug, Clone, Eq, Hash, PartialEq)]
+pub struct PeerIdSerde(pub(crate) PeerId);
+
+impl From<PeerId> for PeerIdSerde {
+    fn from(peer_id: PeerId) -> PeerIdSerde { PeerIdSerde(peer_id) }
+}
+
+impl Serialize for PeerIdSerde {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        self.0.clone().to_bytes().serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for PeerIdSerde {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        let bytes: Vec<u8> = Deserialize::deserialize(deserializer)?;
+        let peer_id = PeerId::from_bytes(&bytes).map_err(|_| serde::de::Error::custom("PeerId::from_bytes error"))?;
+        Ok(PeerIdSerde(peer_id))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum PeersExchangeProtocol {
+    Version1,
+}
+
+impl AsRef<str> for PeersExchangeProtocol {
+    fn as_ref(&self) -> &str {
+        match self {
+            PeersExchangeProtocol::Version1 => "/peers-exchange/1",
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum PeersExchangeRequest {
+    GetKnownPeers { num: usize },
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum PeersExchangeResponse {
+    KnownPeers { peers: HashMap<PeerIdSerde, PeerAddresses> },
+}
+
+pub struct PeersExchange {
+    pub(crate) request_response: RequestResponse<PeersExchangeCodec>,
+    known_peers: Vec<PeerId>,
```

Comment:
Is it possible to use `HashSet` here? Will it be more effective?

Reply:
Indeed, it can improve the performance of `forget_peer` slightly. Added to TODOs.